### PR TITLE
[16.0][MIG] account_statement_import_online_ponto: Migration to 16.0

### PR DESCRIPTION
**Pull request by NL66278**
This PR includes the changes from https://github.com/OCA/bank-statement-import/pull/548

If that base PR is merged, this one can be rebased. Conversely when this branch is merged PR 548 can be closed.

### CODE REVIEW

**Review by inafarrate-busman — APPROVED** [review]

File: account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py [hunk 1]
```diff
@@ -0,0 +1,232 @@
...
+                (
+                    statement_date_since,
+                    statement_date_until,
+                    statement_lines,
+                ) = reset_transactions(statement_date_until)
+            statement_lines.append(vals_line)
+        # Handle lines in last statement.
+        self._create_or_update_statement(
+            (statement_lines, {}), statement_date_since, statement_date_until
+        )
```
Comment (astirpe) [inline account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py]: After testing this module with real data, we noticed that a few lines were retrieved from ponto but were not imported in odoo. The reason seems to be that the date of the lines not always falls back into the date range (date_since - date_until) and so they are filtered out. I had difficulties to fix this method so I ended up with rewriting this method and simplifying the code this way:



```python
    def _ponto_store_lines(self, lines):
        """Store transactions retrieved from Ponto in statements.
        """
        lines = sorted(lines, key=itemgetter("transaction_datetime"))

        # Group statement lines by date per period (date range)
        grouped_periods = {}
        for line in lines:
            date_since = line["transaction_datetime"]
            statement_date_since = self._get_statement_date_since(date_since)
            statement_date_until = statement_date_since + self._get_statement_date_step()
            if (statement_date_since, statement_date_until) not in grouped_periods:
                grouped_periods[(statement_date_since, statement_date_until)] = []

            line.pop("transaction_datetime")
            vals_line = self._ponto_get_transaction_vals(line)
            grouped_periods[(statement_date_since, statement_date_until)].append(vals_line)

        # For each period, create or update statement lines
        for period, statement_lines in grouped_periods.items():
            (statement_date_since, statement_date_until) = period

            self._create_or_update_statement(
                (statement_lines, {}), statement_date_since, statement_date_until
            )
```

**Review by astirpe — CHANGES_REQUESTED** [review]

File: account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py (hunk 1 above)
Comment (NL66278) [reply account_statement_import_online_ponto/models/online_bank_statement_provider_ponto.py]: @astirpe Hi, could you make a PR along these lines to my branch? Just to be sure I updated things like intended. Anyway, thank you for your contribution!

**Review by victoralmau — CHANGES_REQUESTED** [review]
Suggestions:

- There is already `account_statement_import_online_migration` PR (https://github.com/OCA/bank-statement-import/pull/548), we should leave this one only with `account_statement_import_online_ponto` changes.
- Unify all changes in the migration commit.
- Update the translations .pot

**Review by ThijsvOers — APPROVED** [review]
I have tested the Module:
1. created MyPonto account
2. added Odoo integration
3. in Odoo Added online bank sync. Important notice: use Clientid from Myponto in field Login in Odoo
4. Tested with Daily, weekly and Monthly statement, all lines were retreived complete (this was an issue before the lastest changes)
5. Field Partner is now also filled in in odoo, this makes the reconcile process better.

**Review by astirpe — APPROVED** [review]

**Comment by patel1008b** [comment]
@NL66278 Any idea when it will be merge into 16.0 ?

**Comment by NL66278** [comment]
@bhaveshselarka I do not have merge powers in this repository, so I do not know.

**Comment by victoralmau** [comment]
`account_statement_import_online` is already merged, can you rebase to check that everything is correct?

**Comment by NL66278** [comment]
@victoralmau I did the rebase and a check. Found one deprecated text in the help for a field. Everything in order now.

**Review by victoralmau — APPROVED** [review]
Code review OK

**Pull request merged**